File: src/csv_parser.c
#include "csv_parser.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>

Experiment* parse_csv(const char *file_path) {
    FILE *file = fopen(file_path, "r");
    if (!file) {
        perror("Error opening CSV file");
        return NULL;
    }

    Experiment *exp = malloc(sizeof(Experiment));
    exp->stimuli = NULL;
    exp->count = 0;

    char line[512];
    while (fgets(line, sizeof(line), file)) {
        if (line[0] == '#' || line[0] == '\n' || line[0] == '\r' || line[0] == ' ') continue;

        exp->stimuli = realloc(exp->stimuli, (exp->count + 1) * sizeof(Stimulus));
        Stimulus *s = &exp->stimuli[exp->count];

        char type_str[16];
        if (sscanf(line, "%" SCNu64 ",%" SCNu64 ",%15[^,],%255[^,\n\r]", &s->timestamp_ms, &s->duration_ms, type_str, s->file_path) == 4) {
            if (strcmp(type_str, "IMAGE") == 0) s->type = STIM_IMAGE;
            else if (strcmp(type_str, "SOUND") == 0) s->type = STIM_SOUND;
            else if (strcmp(type_str, "TEXT") == 0) s->type = STIM_TEXT;
            else s->type = STIM_END;
            exp->count++;
        }
    }

    fclose(file);
    return exp;
}

void free_experiment(Experiment *exp) {
    if (exp) {
        if (exp->stimuli) free(exp->stimuli);
        free(exp);
    }
}

File: go-expe3000/csv_parser.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func LoadExperiment(path string) (*Experiment, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	var stimuli []Stimulus
	for i, record := range records {
		if len(record) < 4 {
			continue // or return error
		}

		timestamp, err := strconv.ParseUint(record[0], 10, 64)
		if err != nil {
			return nil, fmt.Errorf("line %d: invalid timestamp: %v", i+1, err)
		}

		duration, err := strconv.ParseUint(record[1], 10, 64)
		if err != nil {
			return nil, fmt.Errorf("line %d: invalid duration: %v", i+1, err)
		}

		var stype StimType
		switch strings.ToLower(record[2]) {
		case "image":
			stype = StimImage
		case "sound":
			stype = StimSound
		case "text":
			stype = StimText
		default:
			return nil, fmt.Errorf("line %d: unknown stimulus type: %s", i+1, record[2])
		}

		stimuli = append(stimuli, Stimulus{
			TimestampMS: timestamp,
			DurationMS:  duration,
			Type:        stype,
			FilePath:    record[3],
		})
	}

	return &Experiment{Stimuli: stimuli}, nil
}
